quickwit-indexing/src/source/kinesis/kinesis_source.rs

```diff
@@ -0,0 +1,530 @@
+    lag_millis: Option<i64>,
+    shard_consumer_handle: ShardConsumerHandle,
+}
+
+#[derive(Default)]
+pub struct KinesisSourceState {
+    /// Pool of [`ShardConsumer`] managed by the source.
+    shard_consumers: HashMap<ShardId, ShardConsumerState>,
+    /// Number of bytes processed by the source.
+    pub num_bytes_processed: u64,
+    /// Number of records processed by the source (including invalid messages).
+    pub num_records_processed: u64,
+    // Number of invalid records, i.e., that were empty or could not be parsed.
+    pub num_invalid_records: u64,
+}
+
+pub struct KinesisSource {
+    stream_name: String,
+    checkpoint: SourceCheckpoint,
+    kinesis_client: KinesisClient,
+    sink_tx: mpsc::Sender<ShardConsumerMessage>,
+    sink_rx: mpsc::Receiver<ShardConsumerMessage>,
+    state: KinesisSourceState,
+}
+
+impl fmt::Debug for KinesisSource {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "KinesisSource {{ stream_name: {} }}", self.stream_name)
+    }
+}
+
+impl KinesisSource {
+    /// Instantiates a new `KinesisSource`.
+    pub async fn try_new(
+        params: KinesisSourceParams,
+        checkpoint: SourceCheckpoint,
+    ) -> anyhow::Result<Self> {
+        let stream_name = params.stream_name;
+        let region = get_region(params.region, params.endpoint);
+        let kinesis_client = KinesisClient::new(region);
+        let (sink_tx, sink_rx) = mpsc::channel(1_000);
+        let state = KinesisSourceState::default();
+        Ok(KinesisSource {
+            stream_name,
+            checkpoint,
+            kinesis_client,
+            sink_tx,
+            sink_rx,
+            state,
+        })
+    }
+
+    fn spawn_shard_consumer(&mut self, ctx: &SourceContext, shard_id: ShardId) {
+        assert!(!self.state.shard_consumers.contains_key(&shard_id));
+
+        let partition_id = PartitionId::from(shard_id.as_ref());
+        let position = self
+            .checkpoint
+            .position_for_partition(&partition_id)
+            .cloned()
+            .unwrap_or(Position::Beginning);
+        let from_sequence_number_exclusive = match &position {
+            Position::Offset(offset) => Some(offset.to_string()),
+            Position::Beginning => None,
+        };
+        let shard_consumer = ShardConsumer::new(
+            self.stream_name.clone(),
+            shard_id.clone(),
+            from_sequence_number_exclusive,
+            true,
+            self.kinesis_client.clone(),
+            self.sink_tx.clone(),
+        );
+        let shard_consumer_handle = shard_consumer.spawn(ctx);
+        let shard_consumer_state = ShardConsumerState {
+            partition_id,
+            position,
+            lag_millis: None,
+            shard_consumer_handle,
+        };
+        self.state
+            .shard_consumers
+            .insert(shard_id, shard_consumer_state);
+    }
+
+    fn supervise_shard_consumers(&mut self) {
+        // TODO
```

Comment:
What kind of retry algorithm do we want here? We can start with something as simple as failing the source, which the indexing pipeline will respawn.

Reply:
sure